src/lib/promise/native_helpers/promise_native_helpers.ml

```diff
@@ -0,0 +1,7 @@
+(* there is no risk in doing Obj.magic here -- Promise.t is the same as Deferred.t in the native build *)
```

Comment:
There's nothing I can think of off the top of my head, but is there any way we can cause this to fail if someone changes Promise without realizing we rely on this? ie will a bunch of tests just segfault or will things break in more subtle ways and we would potentially not notice.

Reply:
I'm sure you're already doing it, but you can easily implement `of_deferred` if you add a `val create : (('a -> unit) -> unit) -> 'a t` with something like
```ocaml
let create f =
  let res = Ivar.create () in
  f (Ivar.fill_if_empty res);
  Ivar.read res
```

```js
function create(f) {
  var newDeferred = {
    promise: new Promise(function (resolve, error) {
        var true_resolve = function(res) {
          newDeferred.value = res;
          newDeferred.isDetermined = true;
        };
        try {
          f(true_resolve);
        } catch(e) {
          newDeferred.error = e;
          newDeferred.isDetermined = true;
          error(e);
        }
      }),
    isDetermined: false,
  };
  return newDeferred;
}
```
and then
```ocaml
let of_deferred d = Promise.create (fun fill -> Deferred.upon d fill)
```

Reply:
cool, I'll add that in as well!

Reply:
Thinking some more: `of_deferred` is probably a foot-gun that is best avoided. For example, if you were to run some javascript code without a scheduler, an `of_deferred` will give you a promise, but that promise will never resolve because the job will never be scheduled.

We could rename it to `of_deferred_unsafe` with a comment that a scheduler must be running, but it might be safer to remove it entirely if you don't have a specific use for it.

Reply:
Well, OK let me remove it again :sweat_smile:

Reply:
Done!